### PR TITLE
Pull request/0437e811

### PR DESCRIPTION
**Pull request by eldadHaber**
Added circular cinvolution

### CODE REVIEW

File: src/kernelTypes/convCircKernel.jl [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+end
+
+function nImgOut(this::convCircKernel)
+    return [this.nImg[1]; this.nImg[2]; this.sK[3]]
+ end
+
+function getConvCircKernel(TYPE::Type,nImg,sK)
+ 
+    n3 = [nImg[1],nImg[2],sK[3]]
+    Ind = zeros(n3[1],n3[2],n3[3])
```
Comment (klensink) [inline src/kernelTypes/convCircKernel.jl:17]: If `n3` ever has more elements and you're tired of writing this out, `zeros(n3...)` does the same thing. It's called 'splatting'

File: src/kernelTypes/convCircKernel.jl [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    nImg :: Array{Int,1}
+    sK   :: Array{Int,1}
+    iK   :: Array{Int,1}
+end
+
+function nImgOut(this::convCircKernel)
+    return [this.nImg[1]; this.nImg[2]; this.sK[3]]
+ end
+
+function getConvCircKernel(TYPE::Type,nImg,sK)
```
Comment (klensink) [inline src/kernelTypes/convCircKernel.jl:14]: Just FYI outer constructors are usually named the same as the type, but this `getType` style is done throughout the package so we can just leave it for now

File: src/kernelTypes/convCircKernel.jl [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+
+export Amv
+function Amv(this::convCircKernel{T},theta::Array{T},Y::Array{T}) where {T<:Number}
+
+    nex   = div(length(Y),prod(nImgIn(this)))
+
+    n     = nImgIn(this);
+    
+    Y     = reshape(Y,(n[1],n[2],this.sK[3],nex))
```
Comment (klensink) [inline src/kernelTypes/convCircKernel.jl]: Type instability,

Please change the input in line 32 to `Y_in::Array{T}` and this line to `Y     = reshape(Y_in, (n[1], n[2], this.sK[3], nex))`

File: src/kernelTypes/convCircKernel.jl [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    Z  = real(ifft(Kh.*Yh,(3,2,1)))
+    return Z             
+end
+
+export ATmv
+function ATmv(this::convCircKernel{T},theta::Array{T},Z::Array{T}) where {T<:Number}
+    
+    nex   =  div(numel(Z),prod(nImgIn(this)))
+    n     = nImgIn(this)
+    Z     = reshape(Z,(n[1],n[2],this.sK[3],nex));
```
Comment (klensink) [inline src/kernelTypes/convCircKernel.jl]: Type instability

See comment above

File: src/kernelTypes/convCircKernel.jl [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    K[this.iK] = theta
+    
+    Kh = fft(K,(1,2,3))
+    Zh = fft(Z,(1,2,3))
+    
+    Y  = real(ifft(conj(Kh).*Zh,(3,2,1)))
+    return Y
+end
+
+function Jthetamv(this::convCircKernel{T},dtheta::Array{T},dummy::Array{T},Y::Array{T},temp=nothing) where {T<:Number}
```
Comment (klensink) [inline src/kernelTypes/convCircKernel.jl]: Is `dummy` just to keep the function call to `Jthetamv` the same as other types that need that input?

File: src/kernelTypes/convCircKernel.jl [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+    nex    =  div(numel(Y),nFeatIn(this))
+    Y      = reshape(Y,:,nex)
```
Comment (klensink) [inline src/kernelTypes/convCircKernel.jl]: Type instability

File: src/kernelTypes/convCircKernel.jl [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+end
+
+function JthetaTmv(this::convCircKernel{T},Z::Array{T},dummy::Array{T},Y::Array{T}) where {T<:Number}
+    #  derivative of Z*(A(theta)*Y) w.r.t. theta
+
+    nex  =  div(numel(Y),nFeatIn(this))
+
+    n  = nImgIn(this)
+    
+    Z     = reshape(Z,n[1],n[2],this.sK[3],nex)
```
Comment (klensink) [inline src/kernelTypes/convCircKernel.jl]: Type instability

File: src/kernelTypes/convCircKernel.jl [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    Y     = reshape(Y,n[1],n[2],this.sK[3],nex)
```
Comment (klensink) [inline src/kernelTypes/convCircKernel.jl]: Type instability

**Review by klensink — APPROVED** [review]

File: src/kernelTypes/convCircKernel.jl (hunk 1 above)
Comment (lruthotto) [reply src/kernelTypes/convCircKernel.jl:17]: cool. good to know. it that the same as `tuple(n3...)` ?

Comment (klensink) [reply src/kernelTypes/convCircKernel.jl:17]: Yup. It only works for function calls, but it just passes each element of `n3` as a different argument.

The opposite can be done in function definitions, and is called 'slurping'. You can collect any number of input arguments into a tuple. For example below I'll slurp up the variable input then splat it into the addition operator.

```
julia> function foo(x...)
           println(x)
           +(x...)
       end
foo (generic function with 1 method)

julia> foo(1,2,3)
(1, 2, 3)
6

julia> foo(1,2,3,4,5,6)
(1, 2, 3, 4, 5, 6)
21

julia> foo(1,2,3,4,5,6,3,4,5,6,5,3,34,4)
(1, 2, 3, 4, 5, 6, 3, 4, 5, 6, 5, 3, 34, 4)
85
```

**Pull request merged**